File: room.py
from globals import *

__all__ = ['light_up_room', 'light_passage', 'darken_room', 'get_room_char',
           'get_rand_row_col', 'get_rand_room', 'fill_room_with_objects',
           'get_room_number', 'shell']

def light_up_room():
    if g.blind: return
    r = rooms[g.current_room]
    for i in range(r.top_row, r.bottom_row + 1):
        for j in range(r.left_col, r.right_col + 1):
            mvaddch(i, j, get_room_char(screen[i][j], i, j))
    mvaddch(rogue.row, rogue.col, rogue.fchar)

def light_passage(row, col):
    if g.blind: return
    i_end = 1 if row < LINES - 2 else 0
    j_end = 1 if col < COLS - 1 else 0
    
    for i in range(-1 if row > MIN_ROW else 0, i_end + 1):
        for j in range(-1 if col > 0 else 0, j_end + 1):
            if is_passable(row + i, col + j):
                r = row + i
                c = col + j
                mvaddch(r, c, get_room_char(screen[r][c], r, c))

def darken_room(rn):
    if g.blind: return
    r = rooms[rn]
    for i in range(r.top_row + 1, r.bottom_row):
        for j in range(r.left_col + 1, r.right_col):
            if not is_object(i, j) and not (g.detect_monster and screen[i][j] & MONSTER):
                if not hiding_xeroc(i, j):
                    mvaddch(i, j, ' ')

def get_room_char(mask, row, col):
    if mask & MONSTER:
        return get_monster_char_row_col(row, col)
    if mask & SCROLL:
        return '?'
    if mask & POTION:
        return '!'
    if mask & FOOD:
        return ':'
    if mask & WAND:
        return '/'
    if mask & ARMOR:
        return ']'
    if mask & WEAPON:
        return ')'
    if mask & GOLD:
        return '*'
    if mask & TUNNEL:
        return '#'
    if mask & HORWALL:
        return '-'
    if mask & VERTWALL:
        return '|'
    if mask & AMULET:
        return ','
    if mask & FLOOR:
        return '.'
    if mask & DOOR:
        return '+'
    if mask & STAIRS:
        return '%'
    return ' '

def get_rand_row_col(mask):
    while True:
        row = get_rand(MIN_ROW, SROWS - 2)
        col = get_rand(0, SCOLS - 1)
        rn = get_room_number(row, col)
        if screen[row][col] & mask and not screen[row][col] & ~mask and rn != NO_ROOM: break
    return row, col

def get_rand_room():
    while True:
        i = get_rand(0, MAXROOMS - 1)
        if rooms[i].is_room: break
    return i

def fill_room_with_objects(rn):
    r = rooms[rn]
    N = (r.bottom_row - r.top_row - 1) * (r.right_col - r.left_col - 1)
    n = get_rand(5, 10)
    if n > N: n = N - 2
    
    for i in range(n):
        while True:
            row = get_rand(r.top_row + 1, r.bottom_row - 1)
            col = get_rand(r.left_col + 1, r.right_col - 1)
            if screen[row][col] == FLOOR: break
        obj = get_rand_object()
        put_object_at(obj, row, col)
        
    return n

def get_room_number(row, col):
    for i in range(MAXROOMS):
        r = rooms[i]
        if r.top_row <= row <= r.bottom_row and r.left_col <= col <= r.right_col:
            return i
    return NO_ROOM

def shell():
    raise

File: flrogue/lib/rogue/room.dart
import 'globals.dart';
import 'monster.dart';
import 'move.dart';
import 'special_hit.dart';
import 'ui.dart';

void lightUpRoom() {
  if (blind != 0) return;

  Room r = rooms[currentRoom];

  for (int i = r.topRow; i <= r.bottomRow; i++) {
    for (int j = r.leftCol; j <= r.rightCol; j++) {
      ui.move(i, j).write(getRoomChar(screen[i][j], i, j));
    }
  }

  ui.move(rogue.row, rogue.col).write(rogue.fchar);
}

void lightPassage(int row, int col) {
  if (blind != 0) return;

  int iEnd = row < ui.rows - 2 ? 1 : 0;
  int jEnd = col < ui.cols - 1 ? 1 : 0;

  for (int i = (row > minRow ? -1 : 0); i <= iEnd; i++) {
    for (int j = (col > 0 ? -1 : 0); j <= jEnd; j++) {
      if (isPassable(row + i, col + j)) {
        int r = row + i;
        int c = col + j;
        ui.move(r, c).write(getRoomChar(screen[r][c], r, c));
      }
    }
  }
}

void darkenRoom(int rn) {
  if (blind != 0) return;

  Room r = rooms[rn];

  for (int i = r.topRow + 1; i < r.bottomRow; i++) {
    for (int j = r.leftCol + 1; j < r.rightCol; j++) {
      if (!isObject(i, j) &&
          !(detectMonster && screen[i][j] & Cell.monster != 0)) {
        if (!hidingXeroc(i, j)) {
          ui.move(i, j).write(' ');
        }
      }
    }
  }
}

String getRoomChar(int mask, int row, int col) {
  if (mask & Cell.monster != 0) {
    return getMonsterCharRowCol(row, col);
  }

  if (mask & Cell.scroll != 0) {
    return '?';
  }

  if (mask & Cell.potion != 0) {
    return '!';
  }

  if (mask & Cell.food != 0) {
    return ':';
  }

  if (mask & Cell.wand != 0) {
    return '/';
  }

  if (mask & Cell.armor != 0) {
    return ']';
  }

  if (mask & Cell.weapon != 0) {
    return ')';
  }

  if (mask & Cell.gold != 0) {
    return '*';
  }

  if (mask & Cell.tunnel != 0) {
    return '#';
  }

  if (mask & Cell.horWall != 0) {
    return '-';
  }

  if (mask & Cell.vertWall != 0) {
    return '|';
  }

  if (mask & Cell.amulet != 0) {
    return ',';
  }

  if (mask & Cell.stairs != 0) {
    return '%';
  }

  if (mask & Cell.floor != 0) {
    return '.';
  }

  if (mask & Cell.door != 0) {
    return '+';
  }

  return ' ';
}

(int, int) getRandRowCol(int mask) {
  int row, col;

  while (true) {
    row = getRand(minRow, sRows - 2);
    col = getRand(0, sCols - 1);
    int rn = getRoomNumber(row, col);
    if (screen[row][col] & mask != 0 &&
        screen[row][col] & ~mask == 0 &&
        rn != noRoom) {
      break;
    }
  }

  return (row, col);
}

int getRandRoom() {
  int i;

  while (true) {
    i = getRand(0, maxRooms - 1);
    if (rooms[i].isRoom) break;
  }

  return i;
}

int getRoomNumber(int row, int col) {
  for (int i = 0; i < maxRooms; i++) {
    Room r = rooms[i];
    if (r.topRow <= row &&
        row <= r.bottomRow &&
        r.leftCol <= col &&
        col <= r.rightCol) {
      return i;
    }
  }
  return noRoom;
}

void drawMagicMap() {
  const mask =
      Cell.horWall | Cell.vertWall | Cell.door | Cell.tunnel | Cell.stairs;

  for (var i = 0; i < sRows; i++) {
    for (var j = 0; j < sCols; j++) {
      var s = screen[i][j];
      if (s & mask != 0) {
        var ch = ui.move(i, j).read();
        if (ch == ' ') {
          if (s & Cell.horWall != 0) ch = '-';
          if (s & Cell.vertWall != 0) ch = '|';
          if (s & Cell.door != 0) ch = '+';
          if (s & Cell.tunnel != 0) ch = '#';
          if (s & Cell.stairs != 0) ch = '%';
          ui.move(i, j).write(ch);
        }
      }
    }
  }
}
